feat: add how-to for multiple architecture (#395)
oras-project/oras-www: versioned_docs/version-1.3.0-beta.1/how_to_guides/multiarch.mdx

---

> title: Multiple Architecture Image Support

Reviewer: ```suggestion
title: Multi-architecture image/artifact support
```

Reviewer: Maybe we can follow the common term `multi-architecture` from docker: https://www.docker.com/blog/tag/multi-architecture/

Reviewer 2: `Multi-architecture Image / Artifact Support` <- You may want to capitalize each word.

Author: Done.

sidebar_position: 110
---

> # Multiple Architecture Image/Artifact Support with ORAS

Reviewer: ```suggestion
# Multi-architecture image/artifact support
```

Author: Done

> ORAS can help you manage multiple architecture artifacts such as configurations, executables, and container images.

Reviewer: ```suggestion
ORAS can help you create and manage multi-architecture artifacts such as configurations, executables, and container images.
```

This example shows how you can use ORAS to manage executables for different operating systems and architectures.

## Pushing the artifacts

> The first step is to push the artifacts to your container registry.

Reviewer: ```suggestion
The first step is to push the artifact with platform information to your container registry.
```

First we push a Linux ARM64 executable:

> ```shell

Reviewer 2: consider
```suggestion
```console
```

> % oras push --plain-http localhost:15000/oras:linux,arm64  --artifact-platform linux/arm64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/linux/arm64/oras:application/octet-stream

Reviewer 2: nit: you may use `$` instead `%`.

Author: I'd sooner use nothing so copy/paste is easier come to think of it.

Author: Changed my mind. The way this document is with command and output, I'll go with `$`.

Done.

> % oras push --plain-http localhost:15000/oras:linux,arm64  --artifact-platform linux/arm64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/linux/arm64/oras:application/octet-stream

Reviewer 2: nit: remove extra spaces
```suggestion
% oras push --plain-http localhost:15000/oras:linux,arm64 --artifact-platform linux/arm64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/linux/arm64/oras:application/octet-stream
```

```

Next, we push a Linux AMD64 executable:
```shell
% oras push --plain-http localhost:15000/oras:linux,amd64  --artifact-platform linux/amd64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/linux/amd64/oras:application/octet-stream
✓ Uploaded  bin/linux/amd64/oras                                                                                                         11/11 MB 100.00%     0s
  └─ sha256:4640fefdb4e22eb1308e40d5fac9fd9d16e17dfbc20202bb3ed42033d01d823e
✓ Exists    application/vnd.unknown.config.v1+json                                                                                       37/37  B 100.00%     0s
  └─ sha256:9d99a75171aea000c711b34c0e5e3f28d3d537dd99d110eafbfbc2bd8e52c2bf
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                                                 581/581  B 100.00%     0s
  └─ sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
Pushed [registry] localhost:15000/oras:linux
Tagged amd64
ArtifactType: application/vnd.oci.image.config.v1+json
Digest: sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
```

> This example pushes the executable for two architectures, but you could also push artifacts with a different operating system or other platform attribute.

Reviewer 2: Question: What's the meaning to tag `linux`?

Reviewer 2: or do you mean `linux_arm64` and `linux_amd64`?

Author: Fixed

> ## Create a multiple architecture manifest

Reviewer: ```suggestion
## Create a multi-architecture manifest
```

Author: Done.

> After you have pushed all your platform specific artifacts, create a manifest referencing them:

Reviewer: ```suggestion
After you have pushed all your platform specific artifacts, create a manifest referencing them using the OCI image index format:
```

Reviewer: It worth telling users the format of the created manifest

```shell

> % oras manifest index create localhost:15000/oras:v1 sha256:302df4334cffe5132b6b81c06742be8ab3345bff3ce922662bc8674e79113bfd sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4

Reviewer 2: You may also want to have a section to create index using tags.

Author: Done.

Fetching  sha256:302df4334cffe5132b6b81c06742be8ab3345bff3ce922662bc8674e79113bfd
Fetched   sha256:302df4334cffe5132b6b81c06742be8ab3345bff3ce922662bc8674e79113bfd
Fetching  sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
Fetched   sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
Packed    0fd4ac889c6f application/vnd.oci.image.index.v1+json
Pushed    [registry] localhost:15000/oras:v1
Digest: sha256:0fd4ac889c6f1f3e4fbb1c88b0cc18b97176bf7fd0841ac9c5264d20dd54e2d8
```

> 

Reviewer: I would suggest adding a section "View the multi-architecture manifest" and using `oras manifest fetch localhost:15000/oras:v1` to show the manifest JSON metadata. It can help users to list the created manifest list with different platforms.

Author: Done.

---
title: Multi-Architecture Image/Artifact Support
sidebar_position: 110
---

# Multi-Architecture Image/Artifact Support

ORAS can help you create and manage multi-architecture artifacts such as configurations, executables, and container images.
This example shows how you can use ORAS to manage executables for different operating systems and architectures.

## Pushing the artifacts
The first step is to push the artifact with platform information to your container registry.

First we push a Linux ARM64 executable:
```console

> $ oras push --plain-http localhost:15000/oras:linux_arm64 --artifact-platform linux/arm64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/linux/arm64/oras:application/octet-stream

Reviewer 3: Do we want to put specify a more meaningful artifact type? The artifact to push is not an image config. 

BTW, Is the `--plain-http` flag necessary here? I feel that it makes the command too long.

Author: Anyone have thoughts on what would be a better artifact-type here? I thought it was an image config containing `application/octet-stream`.

Reviewer 3: Should the artifact type refer to the type of `bin/linux/arm64/oras`?

Author: Got rid of `--plain-http` I don't think artificate type should refer to `application/octet-stream` that is the type of the artifact itself

Reviewer 3: How about something like `"application/vnd.example` 😂

  └─ sha256:4640fefdb4e22eb1308e40d5fac9fd9d16e17dfbc20202bb3ed42033d01d823e
✓ Exists    application/vnd.unknown.config.v1+json                                                                                       37/37  B 100.00%     0s
  └─ sha256:9d99a75171aea000c711b34c0e5e3f28d3d537dd99d110eafbfbc2bd8e52c2bf
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                                                 581/581  B 100.00%     0s
  └─ sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
Pushed [registry] localhost:15000/oras:linux_amd64
Tagged amd64
ArtifactType: application/vnd.oci.image.config.v1+json
Digest: sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4
```
This example pushes the executable for two architectures, but you could also push artifacts with a different operating system or other platform attribute.

## Create a multi-architecture manifest

After you have pushed all your platform specific artifacts, create a manifest referencing them using the OCI image index format:
```console

> $ oras manifest index create localhost:15000/oras:v1 sha256:302df4334cffe5132b6b81c06742be8ab3345bff3ce922662bc8674e79113bfd sha256:e5b0ffe309a91893d4423f09e82410364607fd749ea2f93dd1c286d718d9c1e4

Reviewer 3: nit: How about putting the digests into environment variables for better readability?

Author: Done

```console
$ oras push localhost:15000/oras:darwin,amd64  --artifact-platform darwin/amd64 --artifact-type 'application/vnd.oci.image.config.v1+json'  bin/darwin/amd64/oras:application/octet-stream
✓ Exists    bin/darwin/amd64/oras                                                                                                    11.1/11.1 MB 100.00%     0s
  └─ sha256:a839f5d9cd3502b87b8d8edca31cd9f6f66c4835364e73b87a1645353631590f
✓ Exists    application/vnd.unknown.config.v1+json                                                                                       38/38  B 100.00%     0s
  └─ sha256:e9302bbb2fb8f6c2df866d3c4e41917849442f89a575f36f43366a7279804f70
✓ Uploaded  application/vnd.oci.image.manifest.v1+json                                                                                 589/589  B 100.00%     0s
  └─ sha256:f5758b75ebc4d82280078afb9ebeb859f91a481667c018c9edf91a432518cb20
Pushed [registry] localhost:15000/oras:darwin
Tagged amd64
ArtifactType: application/vnd.oci.image.config.v1+json
Digest: sha256:f5758b75ebc4d82280078afb9ebeb859f91a481667c018c9edf91a432518cb20
````

Add it to your manifest:
```console

> $ oras manifest index update localhost:15000/oras:v1 --add sha256:f5758b75ebc4d82280078afb9ebeb859f91a481667c018c9edf91a432518cb20

Reviewer 3: Use an environment variable for the digest here as well?
